contracts/collections/sg721-base/src/contract.rs

```diff
@@ -1,275 +1,172 @@
-use cw721_base::state::TokenInfo;
-use url::Url;
+use cw721::msg::{CollectionInfoMsg, Cw721MigrateMsg};
+use cw721::state::{MAX_COLLECTION_DESCRIPTION_LENGTH, MINTER};
+use cw721::traits::{Contains, Cw721CustomMsg, Cw721State};
+use cw721_base::msg::{CollectionExtensionMsg, RoyaltyInfoResponse};
+use cw721_base::{
+    traits::StateFactory, DefaultOptionalCollectionExtension, DefaultOptionalCollectionExtensionMsg,
+};
 
 use cosmwasm_std::{
-    to_json_binary, Addr, Binary, ContractInfoResponse, Decimal, Deps, DepsMut, Empty, Env, Event,
-    MessageInfo, Response, StdError, StdResult, Storage, Timestamp, WasmQuery,
+    to_json_binary, Addr, Binary, ContractInfoResponse, CustomMsg, Deps, DepsMut, Env, Event,
+    MessageInfo, Response, StdError, Storage, Timestamp, WasmQuery,
 };
 
-use cw721::{ContractInfoResponse as CW721ContractInfoResponse, Cw721Execute};
+use cw721_base::traits::{Cw721Execute, Cw721Query};
 use cw_utils::nonpayable;
-use serde::{de::DeserializeOwned, Serialize};
 
-use sg721::{
-    CollectionInfo, ExecuteMsg, InstantiateMsg, RoyaltyInfo, RoyaltyInfoResponse,
-    UpdateCollectionInfoMsg,
-};
+#[allow(deprecated)]
+use sg721::{ExecuteMsg, InstantiateMsg, UpdateCollectionInfoMsg};
 
+#[allow(deprecated)]
 use crate::msg::{CollectionInfoResponse, NftParams, QueryMsg};
 use crate::{ContractError, Sg721Contract};
 
 use crate::entry::{CONTRACT_NAME, CONTRACT_VERSION};
 
-const MAX_DESCRIPTION_LENGTH: u32 = 512;
-const MAX_SHARE_DELTA_PCT: u64 = 2;
-const MAX_ROYALTY_SHARE_PCT: u64 = 10;
-
-impl<'a, T> Sg721Contract<'a, T>
+impl<
+        'a,
+        TNftExtension,
+        TNftExtensionMsg,
+        TExtensionMsg,
+        TExtensionQueryMsg,
+        TCustomResponseMsg,
+    >
+    Sg721Contract<
+        'a,
+        // NftInfo extension (onchain metadata).
+        TNftExtension,
+        // NftInfo extension msg for onchain metadata.
+        TNftExtensionMsg,
```

Comment:
RE naming conventions. I'm just curious if we need to name variables this way. The comment says NFTInfo extension msg for onchain metadata. Would ie be wrong to name it NFTInfoOnchainMetadataExtensionMsg or OnchainMetadataExtensionMsg something like that? I see this pattern in a lot of variables where the description is different than the name. Or do we need these names for consistency?

Reply:
We need this. `T` is for generic. If it would be e.g. just `NftExtension` than it is a specific struct. Look how `Sg721Contract` is used:

```rs
        Sg721Contract::<
            DefaultOptionalNftExtension,
            DefaultOptionalNftExtensionMsg,
            DefaultOptionalCollectionExtension,
            DefaultOptionalCollectionExtensionMsg,
            Empty,
            Empty,
            Empty,
        >::default()
        .execute(deps, env, info, msg)
```

Now if Stargaze lateron wants to provide another custom contract sg721, and provide a different type like MyCustomNftExtension for `TNftExtension`. Then it may look like this:

```rs
        Sg721Contract::<
            MyCustomNftExtension,
            MyCustomNftExtensionMsg, // optional custom msg, but doesnt has to
            DefaultOptionalCollectionExtension,
            DefaultOptionalCollectionExtensionMsg,
            Empty,
            Empty,
            Empty,
        >::default()
        .execute(deps, env, info, msg)
```

Reply:
> RE naming conventions. I'm just curious if we need to name variables this way. The comment says NFTInfo extension msg for onchain metadata. Would ie be wrong to name it NFTInfoOnchainMetadataExtensionMsg or OnchainMetadataExtensionMsg something like that? I see this pattern in a lot of variables where the description is different than the name. Or do we need these names for consistency?

I should need change comment, since custom contracts can use this for any logic and type - not only for onchain metadata. This is also the reason why I used `extension` (instead of metadata) as naming for all structs, methods, generics, etc.